linalg/vector.cpp

```diff
@@ -932,12 +932,100 @@ static double cuVectorDot(const int N, const double *X, const double *Y)
 }
 #endif // MFEM_USE_CUDA
 
+#ifdef MFEM_USE_HIP
+static __global__ void hipKernelMin(const int N, double *gdsr, const double *x)
+{
+   __shared__ double s_min[MFEM_CUDA_BLOCKS];
+   const int n = hipBlockDim_x*hipBlockIdx_x + hipThreadIdx_x;
+   if (n>=N) { return; }
+   const int bid = hipBlockIdx_x;
+   const int tid = hipThreadIdx_x;
+   const int bbd = bid*hipBlockDim_x;
+   const int rid = bbd+tid;
+   s_min[tid] = x[n];
+   for (int workers=hipBlockDim_x>>1; workers>0; workers>>=1)
+   {
+      __syncthreads();
+      if (tid >= workers) { continue; }
+      if (rid >= N) { continue; }
+      const int dualTid = tid + workers;
+      if (dualTid >= N) { continue; }
+      const int rdd = bbd+dualTid;
+      if (rdd >= N) { continue; }
+      if (dualTid >= hipBlockDim_x) { continue; }
+      s_min[tid] = fmin(s_min[tid], s_min[dualTid]);
+   }
+   if (tid==0) { gdsr[bid] = s_min[0]; }
+}
+
+static Array<double> cuda_reduce_buf;
+
+static double hipVectorMin(const int N, const double *X)
+{
+   const int tpb = MFEM_CUDA_BLOCKS;
+   const int blockSize = MFEM_CUDA_BLOCKS;
+   const int gridSize = (N+blockSize-1)/blockSize;
+   const int min_sz = (N%tpb)==0? (N/tpb) : (1+N/tpb);
+   cuda_reduce_buf.SetSize(min_sz);
+   Memory<double> &buf = cuda_reduce_buf.GetMemory();
+   double *d_min = buf.Write(MemoryClass::CUDA, min_sz);
```

Comment:
Should the memory class be CUDA?

Reply:
It's still CUDA, in the current memory manager, we don't really have a HIP backend and we are using the CUDA enum as some 'DEVICE' one. All the cuda* are just forwarded to the hip* ones.